voicevox_engine/user_dict.py

```diff
@@ -53,61 +52,80 @@ def update_dict(
-                    accent_type=word.accent_type,
-                    mora_count=word.mora_count,
-                    accent_associative_rule=word.accent_associative_rule,
-                )
+            csv_text += (
+                "{surface},{context_id},{context_id},{cost},{part_of_speech},"
+                + "{part_of_speech_detail_1},{part_of_speech_detail_2},"
+                + "{part_of_speech_detail_3},{inflectional_type},"
+                + "{inflectional_form},{stem},{yomi},{pronunciation},"
+                + "{accent_type}/{mora_count},{accent_associative_rule}\n"
+            ).format(
+                surface=word.surface,
+                context_id=word.context_id,
+                cost=priority2cost(word.context_id, word.priority),
+                part_of_speech=word.part_of_speech,
+                part_of_speech_detail_1=word.part_of_speech_detail_1,
+                part_of_speech_detail_2=word.part_of_speech_detail_2,
+                part_of_speech_detail_3=word.part_of_speech_detail_3,
+                inflectional_type=word.inflectional_type,
+                inflectional_form=word.inflectional_form,
+                stem=word.stem,
+                yomi=word.yomi,
+                pronunciation=word.pronunciation,
+                accent_type=word.accent_type,
+                mora_count=word.mora_count,
+                accent_associative_rule=word.accent_associative_rule,
             )
-    tmp_dict_path = Path(NamedTemporaryFile(delete=False).name).resolve()
-    pyopenjtalk.create_user_dict(
-        str(Path(f.name).resolve(strict=True)),
-        str(tmp_dict_path),
-    )
-    delete_file(f.name)
-    if not tmp_dict_path.is_file():
-        raise RuntimeError("辞書のコンパイル時にエラーが発生しました。")
-    pyopenjtalk.unset_user_dict()
-    try:
-        shutil.move(tmp_dict_path, compiled_dict_path)  # ドライブを跨ぐためPath.replaceが使えない
-    except OSError:
-        traceback.print_exc()
-        if tmp_dict_path.exists():
-            delete_file(tmp_dict_path.name)
-    finally:
+        tmp_csv_path.write_text(csv_text, encoding="utf-8")
+
+        # 辞書.csvをOpenJTalk用にコンパイル
+        pyopenjtalk.create_user_dict(str(tmp_csv_path), str(tmp_compiled_path))
+        if not tmp_compiled_path.is_file():
+            raise RuntimeError("辞書のコンパイル時にエラーが発生しました。")
+
+        # コンパイル済み辞書の置き換え・読み込み
+        pyopenjtalk.unset_user_dict()
+        tmp_compiled_path.replace(compiled_dict_path)
         if compiled_dict_path.is_file():
             pyopenjtalk.set_user_dict(str(compiled_dict_path.resolve(strict=True)))
 
+    except Exception as e:
+        print("Error: Failed to update dictionary.", file=sys.stderr)
+        print(e, file=sys.stderr)
+        raise e
+
+    finally:
+        # 後処理
+        if tmp_csv_path is not None and tmp_csv_path.exists():
+            tmp_csv_path.unlink()
+        if tmp_compiled_path is not None and tmp_compiled_path.exists():
+            tmp_compiled_path.unlink()
```

Comment:
ファイル削除の前にファイルを close した方が良いと思いました（あるいは、`NamedTemporaryFile` 関数を `delete=True` の設定で `with` 構文で使うとファイルの削除処理が不要になりそうでした）。

一応 macOS で手元で同様のコードを実行したところ、意図通り削除されるようではあったのですが、調べたところ close されていないファイルの削除が成功するかどうかは OS 依存で、Windows では失敗するというような記述を見つけました (cf. https://stackoverflow.com/questions/48315926/should-a-file-be-closed-before-removing ) 。

```suggestion
        if tmp_csv_path is not None and tmp_csv_path.exists():
            tmp_csv_file.close()
            tmp_csv_path.unlink()
        if tmp_compiled_path is not None and tmp_compiled_path.exists():
            tmp_compiled_file.close()
            tmp_compiled_path.unlink()
```

（実際に Windows でも元のコードで問題なさそうでしたら、このままでも良いかもしれません。私は以下のようなコードで `./test` ディレクトリ内に作られた一時ファイルが削除されるかどうかを確認しました。macOS では意図どおり削除されました）

```python
from tempfile import NamedTemporaryFile
from pathlib import Path

test_dir = Path("test")
test_dir.mkdir(exist_ok=True)

f = NamedTemporaryFile(mode="w", delete=False, dir=str(test_dir))
fpath = Path(f.name).resolve()
print("fpath:", fpath)

fpath.write_text("hogehoge", encoding="utf-8")

if fpath is not None and fpath.exists():
    # f.close()
    fpath.unlink()
```

Reply:
たしかにwindowsだとcloseしていないのでエラーになりました！
`NamedTemporaryFile`の仕様に付き合わず、自分でtmpファイルを作るのが丸いかなと思ったのでそうしてみます！！